Refactored puzzlebang router + tests

RitamNandi opened the pull request:
Commented out app.use(databaseMiddleware); in app.ts. 
This comes from database-middleware.ts which imports connectToDatabase from utilities.ts. Utilities.ts is still fully Mongo stuff.
Please try to hop into the branch and see if things work as expected

benjuang commented on src/services/puzzlebang/puzzlebang-router.ts:
```diff
@@ -11,25 +11,38 @@ puzzlebangRouter.post(
+        const currentPoints = attendeeData.points ?? 0;
+
+        if (puzzlesCompleted.includes(puzzleId)) {
+            return res.sendStatus(StatusCodes.CONFLICT).send("Puzzle already completed.");
```
Nice, this is definitely a better response than 401

benjuang commented on src/services/puzzlebang/puzzlebang-validators.ts:
```diff
@@ -1,7 +1,7 @@
-    puzzleId: z.string(),
+    userId: z.string(),
+    // email: z.string().email(),
+    puzzleId: z.string().uuid(),
```
FYI: the IDs that PuzzleBang sends over look like `P1`, `P1`..., `B0`, and `MP`.  They're generally short, 2 or 3 character long strings.  They won't look like UUIDs.

RitamNandi replied:
Thanks for calling this out, will just keep it as z.string() and talk to the others about changing the column in the DB

benjuang commented on src/services/puzzlebang/puzzlebang-validators.ts:
```diff
@@ -1,7 +1,7 @@
-    // userId: z.string(),
-    email: z.string().email(),
-    puzzleId: z.string(),
+    userId: z.string(),
```
I don't think PuzzleBang will know the internal userIds for attendees, unless y'all are moving away from emails..?

benjuang commented on src/services/puzzlebang/puzzlebang-router.ts:
```diff
@@ -11,25 +11,38 @@ puzzlebangRouter.post(
+            .single();
+
+        if (error || !attendeeData) {
             return res.sendStatus(StatusCodes.NOT_FOUND);
```
If it's an error, could we return a 500 error, preferably along with the error message?

benjuang commented on src/services/puzzlebang/puzzlebang-router.ts:
```diff
@@ -11,25 +11,38 @@ puzzlebangRouter.post(
+     const { error: updateError } = await SupabaseDB.ATTENDEES.update({
+        puzzles_completed: updatedPuzzles,
+        points: currentPoints + 2,
+    })
```
Fairly non-critical, but wanted to call out that there's a very low probability, but non-zero risk of a race condition here.  If this is called at the same time the points are changed elsewhere, a user could potentially lose points (thread A reads currentPoints on line 27, then some other thread adds points, and then by the time you get to this update, your currentPoints is out of date).

Unfortunately, I'm not familiar enough with SupabaseDB/Node to say for what the best practice would be here.

However, from a higher level DB design perspective (if y'all have the time to refactor), I would actually suggest splitting out the points into a separate table:
```
CREATE TABLE points_earned(
  userId string,
  points number,
  eventId string, /* eventId from events, or "PuzzleBang"? */
  reason string, /* optional, but maybe puzzleId? */
  createdAt datetime,
  PRIMARY KEY(userId, eventId, reason), /* ensures uniqueness of one points row per user + event + reason */
);
```
and then adding together points as necessary.  This also gives you the benefit of being able to show point history and also a better way to do things like a daily scoreboard.

RitamNandi replied:
Will look into this. Another idea I had is that we can use database functions in Supabase, which should also eliminate the race condition

benjuang reviewed (commented):
Hello!  I saw PuzzleBang in the PR name and that caught my attention.  I'm the person who's generally in charge of making sure PuzzleBang.com can talk to the RP API correctly.  I mostly only looked at the router and validator changes and not the test changes - I think my biggest concerns are with the validator changes.  If you want, you can ignore the looooong comment about a race condition - that's more of me being a perfectionist and potentially being nosy.

RitamNandi commented:
@benjuang Thanks for the review! I believe I've addressed all of your suggestions, and I've changed it so that we are validating on puzzle id and email now, and not userId. I'll discuss the race condition issue with the team (currently am leaning towards using a DB function to make it just one operation). Would greatly appreciate if you would take a look at the updates and let me know if you have any other suggestions

benjuang reviewed (approved):
LGTM, thanks for making the updates.
